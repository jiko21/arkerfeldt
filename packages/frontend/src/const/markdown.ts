export const markdownStyle = `
  .flav-md-h {
    margin: 0;
  }
`;
