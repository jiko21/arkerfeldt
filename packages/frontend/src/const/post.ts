export const PublishStatus = {
  PUBLISHED: 'PUBLISHED',
  UNPUBLISHED: 'UNPUBLISHED',
} as const;

export type PublishStatus = typeof PublishStatus[keyof typeof PublishStatus];

export type Post = {
  id: number;
  title: string;
  content: string;
  createdAt: Date;
  updatedAt: Date;
  isDelete: boolean;
  status: PublishStatus;
  authorId: string;
  author: {
    displayName: string;
  };
};

export type PostInputType = {
  title: string;
  content: string;
  status: PublishStatus;
};

export const candidates: { [key in PublishStatus]: string } = {
  [PublishStatus.UNPUBLISHED]: '未公開のまま保存',
  [PublishStatus.PUBLISHED]: '公開する',
};
