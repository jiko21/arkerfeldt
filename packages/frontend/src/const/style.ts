export const Hover = {
  TOP: 500,
  NORMAL: 0,
}