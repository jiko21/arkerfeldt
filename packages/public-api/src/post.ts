import 'reflect-metadata';
import { Field, Int, ObjectType } from '@nestjs/graphql';
import { User } from './user';

@ObjectType()
export class Post {
  @Field(() => Int)
  id: number;

  @Field(() => String)
  title: string;

  @Field(() => String)
  content: string;

  @Field(() => Date)
  createdAt: Date;

  @Field(() => Date)
  updatedAt: Date;

  @Field(() => Boolean)
  isDelete: boolean;

  @Field(() => String)
  status: string;

  @Field(() => User)
  author: User;
}
