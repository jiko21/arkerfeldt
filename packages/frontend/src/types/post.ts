import { PublishStatus } from '@/const/post';

export type Post = {
  id: number;
  title: string;
  content: string;
  createdAt: Date;
  updatedAt: Date;
  isDelete: boolean;
  status: PublishStatus;
  authorId: string;
  author: {
    displayName: string;
  };
};

export type PostResponse = Post & {
  author: {
    displayName: string;
    photoUrl: string;
    uid: string;
  }
};

export type PostInputType = {
  title: string;
  content: string;
  status: PublishStatus;
};
