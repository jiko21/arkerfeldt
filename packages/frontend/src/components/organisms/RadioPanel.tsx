/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import React, { useRef } from 'react';
import { FC, useState } from 'react';
import { Color } from '@/const/color';
import CancelButton from '@/components/atoms/CancelButton';
import EnableButton from '@/components/atoms/EnableButton';
import RadioButtons from '@/components/molecules/RadioButtons';
import { Hover } from '@/const/style';
import { usePanel } from '@/hooks/usePanel';

const buttonStyle = css`
  background-color: ${Color.DISABLE};
  border: none;
  box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
  color: ${Color.ENABLE};
  font-size: 14px;
  font-weight: 700;
  height: 30px;
  min-width: 80px;

  &:hover {
    background-color: ${Color.DISABLE_HOVER};
  }
`;

type ButtonProps = {
  onClick: () => void;
};

const Button: React.FC<ButtonProps> = ({ onClick, children }) => (
  <button css={buttonStyle} onClick={onClick} data-testid='open-btn'>
    {children}
  </button>
);

const containerStyle = css`
  align-items: center;
  background: ${Color.DISABLE};
  box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
  display: flex;
  flex-direction: column;
  height: 180px;
  justify-content: space-around;
  width: 240px;
  position: absolute;

  span {
    color: ${Color.BLACK};
    font-size: 14px;
    font-weight: 700;
  }
`;

const buttonForm = css`
  display: flex;
  flex-direction: column;
  width: ${240 - 32 * 2}px;
`;

const buttonWrapper = css`
  display: flex;
  justify-content: space-between;
  width: ${240 - 32 * 2}px;
`;

const RadioPanelForm: FC<{
  onApply: (value: string) => void;
  toggle: () => void;
  items: string[];
  value: string;
}> = ({ toggle, value, items, onApply }) => {
  const [radioValue, updateRadioValue] = useState(value);
  return (
    <div css={containerStyle}>
      <span>配信ステータス</span>
      <div css={buttonForm}>
        <RadioButtons
          items={items}
          value={radioValue}
          name={'status'}
          onChangeHandler={updateRadioValue}
        />
      </div>
      <div css={buttonWrapper}>
        <CancelButton onClick={toggle} testId='cancel-btn'>
          キャンセル
        </CancelButton>
        <EnableButton
          testId='enable-btn'
          onClick={() => {
            toggle();
            onApply(radioValue);
          }}
        >
          選択
        </EnableButton>
      </div>
    </div>
  );
};

type Props = {
  onApply: (value: string) => void;
  items: string[];
  value: string;
};

const RadioPanel: FC<Props> = ({ onApply, items, value }) => {
  const ref = useRef<HTMLDivElement>(null);
  const { isOpen, toggle } = usePanel(ref);

  return (
    <div ref={ref} css={css`
      z-index: ${Hover.TOP};
    `}>
      <Button onClick={toggle}>status: {value}</Button>
      {isOpen && (<RadioPanelForm onApply={onApply} toggle={toggle} items={items} value={value} />)}
    </div>
  );
};

export default RadioPanel;
