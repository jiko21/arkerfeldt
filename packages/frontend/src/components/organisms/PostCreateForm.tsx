import React, { useMemo, useState } from 'react';
import { TextArea } from '@/components/atoms/TextArea';
import { createFlavMd } from 'flav-md';
import styled from '@emotion/styled';
import { markdownStyle } from '@/const/markdown';
import SelectableButton from '@/components/atoms/SelectableButton';
import { candidates, PublishStatus } from '@/const/post';
import { PostInputType } from '@/types/post';
import TextForm from '@/components/atoms/TextForm';

const Container = styled.div`
  display: flex;
  max-height: 700px;
  margin-bottom: 24px;
`;

const BottomContainer = styled.div`
  float: right;
`;

const DisplayedText = styled.div`
  overflow-x: scroll;
  margin: 0 12px;
  width: 100%;
`;

type Props = {
  initTitle?: string;
  initContent?: string;
  submit: (param: PostInputType) => void;
};

export const PostCreateForm: React.VFC<Props> = ({ submit, initTitle, initContent }) => {
  const [title, setTitle] = useState(initTitle || '');
  const [content, setContent] = useState(initContent || '');
  const flavMd = createFlavMd();
  const updateText = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setContent(e.target.value);
  };

  const updateTitle = (e: React.ChangeEvent<HTMLInputElement>) => {
    setTitle(e.target.value);
  };

  const renderedText = useMemo(() => {
    return flavMd.readMdText(content).readCssText(markdownStyle).build();
  }, [content, flavMd]);

  const onSubmit = (value: PublishStatus) => {
    submit({
      content,
      title,
      status: value,
    });
  };

  return (
    <div>
      <TextForm value={title} onChange={updateTitle} testId="title-text" />
      <Container>
        <TextArea
          value={content}
          onChange={updateText}
          rows={40}
          testId="post-create-form-textarea"
        />
        <DisplayedText dangerouslySetInnerHTML={{ __html: renderedText }} />
      </Container>
      <BottomContainer>
        <SelectableButton<PublishStatus>
          onSubmit={onSubmit}
          candidates={candidates}
          initValue={PublishStatus.UNPUBLISHED}
        />
      </BottomContainer>
    </div>
  );
};
