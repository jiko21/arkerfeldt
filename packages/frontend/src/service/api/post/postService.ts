import { PostInputType } from '@/types/post';
import { postRequest, putRequest } from '@/service/api';

export const createPost = async (post: PostInputType) => {
  await postRequest('/api/v1/posts', {
    ...post,
  });
};

export const updatePost = async (id: number, post: PostInputType) => {
  await putRequest(`/api/v1/posts/${id}`, {
    ...post,
  });
};
