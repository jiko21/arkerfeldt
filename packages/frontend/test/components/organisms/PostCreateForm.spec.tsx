import { PostCreateForm } from '@/components/organisms/PostCreateForm';
import { PublishStatus } from '@/const/post';
import { fireEvent, render, screen } from '@testing-library/react';

describe('PostCreateForm.tsx', () => {
  let submitMock: jest.Mock;
  beforeEach(() => {
    submitMock = jest.fn();
  });

  afterEach(() => {
    submitMock.mockClear();
  });

  it('should correctly render', () => {
    const props = {
      submit: submitMock,
    };
    const { container } = render(<PostCreateForm {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should correctly change input', () => {
    const props = {
      submit: submitMock,
    };
    const { container } = render(<PostCreateForm {...props} />);

    const textarea = screen.getByTestId('post-create-form-textarea');
    fireEvent.change(textarea, {
      target: {
        value: 'aaaaa',
      },
    });

    expect(container).toMatchSnapshot();
  });

  it('should correctly submit', () => {
    const props = {
      submit: submitMock,
    };
    const { container } = render(<PostCreateForm {...props} />);

    const textarea = screen.getByTestId('post-create-form-textarea');
    fireEvent.change(textarea, {
      target: {
        value: 'content',
      },
    });

    const titleText = screen.getByTestId('title-text');
    fireEvent.change(titleText, {
      target: {
        value: 'title',
      },
    });

    const submitButton = screen.getByTestId('button-area');
    fireEvent.click(submitButton);
    expect(submitMock).toBeCalledWith({
      title: 'title',
      content: 'content',
      status: PublishStatus.UNPUBLISHED,
    });
  });
});
