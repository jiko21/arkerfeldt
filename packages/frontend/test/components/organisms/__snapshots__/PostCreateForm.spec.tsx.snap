// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCreateForm.tsx should correctly change input 1`] = `
.emotion-0 {
  border: 1px solid #2d2d2d;
  border-radius: 6px;
  height: 30px;
  min-width: 240px;
  width: 100%;
}

.emotion-0::-webkit-input-placeholder {
  color: #878686;
}

.emotion-0::-moz-placeholder {
  color: #878686;
}

.emotion-0:-ms-input-placeholder {
  color: #878686;
}

.emotion-0::placeholder {
  color: #878686;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  border: 1px solid #2d2d2d;
  border-radius: 6px;
  width: 100%;
}

.emotion-2::-webkit-input-placeholder {
  color: #878686;
}

.emotion-2::-moz-placeholder {
  color: #878686;
}

.emotion-2:-ms-input-placeholder {
  color: #878686;
}

.emotion-2::placeholder {
  color: #878686;
}

.emotion-3 {
  padding: 0 24px;
  width: 100%;
}

.emotion-4 {
  float: right;
}

.emotion-5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #007AAE;
  border: none;
  box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
  color: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 14px;
  font-weight: 700;
  height: 30px;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  min-width: 240px;
  padding: 0;
}

.emotion-5:hover {
  background-color: #005b82;
}

.emotion-6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-right: 1px solid #FFFFFF;
  height: 100%;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 30px;
}

.emotion-7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 100%;
}

<div>
  <div>
    <input
      class="emotion-0"
      data-testid="title-text"
      value=""
    />
    <div
      class="emotion-1"
    >
      <textarea
        class="emotion-2"
        data-testid="post-create-form-textarea"
        rows="40"
      >
        aaaaa
      </textarea>
      <div
        class="emotion-3"
      >
        <style>
          
  .flav-md-h {
    margin: 0;
  }

        </style>
        

        <p
          class="flav-md-text flav-md-p"
        >
          aaaaa
        </p>
      </div>
    </div>
    <div
      class="emotion-4"
    >
      <div>
        <button
          class="emotion-5"
          data-testid="enable-btn"
        >
          <span
            class="emotion-6"
            data-testid="button-tapped"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down fa-w-14 "
              data-icon="chevron-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                fill="currentColor"
              />
            </svg>
          </span>
          <div
            class="emotion-7"
            data-testid="button-area"
          >
            未公開のまま保存
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`PostCreateForm.tsx should correctly render 1`] = `
.emotion-0 {
  border: 1px solid #2d2d2d;
  border-radius: 6px;
  height: 30px;
  min-width: 240px;
  width: 100%;
}

.emotion-0::-webkit-input-placeholder {
  color: #878686;
}

.emotion-0::-moz-placeholder {
  color: #878686;
}

.emotion-0:-ms-input-placeholder {
  color: #878686;
}

.emotion-0::placeholder {
  color: #878686;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  border: 1px solid #2d2d2d;
  border-radius: 6px;
  width: 100%;
}

.emotion-2::-webkit-input-placeholder {
  color: #878686;
}

.emotion-2::-moz-placeholder {
  color: #878686;
}

.emotion-2:-ms-input-placeholder {
  color: #878686;
}

.emotion-2::placeholder {
  color: #878686;
}

.emotion-3 {
  padding: 0 24px;
  width: 100%;
}

.emotion-4 {
  float: right;
}

.emotion-5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #007AAE;
  border: none;
  box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
  color: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 14px;
  font-weight: 700;
  height: 30px;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  min-width: 240px;
  padding: 0;
}

.emotion-5:hover {
  background-color: #005b82;
}

.emotion-6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-right: 1px solid #FFFFFF;
  height: 100%;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 30px;
}

.emotion-7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 100%;
}

<div>
  <div>
    <input
      class="emotion-0"
      data-testid="title-text"
      value=""
    />
    <div
      class="emotion-1"
    >
      <textarea
        class="emotion-2"
        data-testid="post-create-form-textarea"
        rows="40"
      />
      <div
        class="emotion-3"
      >
        <style>
          
  .flav-md-h {
    margin: 0;
  }

        </style>
        

        <p
          class="flav-md-text flav-md-p"
        />
      </div>
    </div>
    <div
      class="emotion-4"
    >
      <div>
        <button
          class="emotion-5"
          data-testid="enable-btn"
        >
          <span
            class="emotion-6"
            data-testid="button-tapped"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-down fa-w-14 "
              data-icon="chevron-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                fill="currentColor"
              />
            </svg>
          </span>
          <div
            class="emotion-7"
            data-testid="button-area"
          >
            未公開のまま保存
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
